floors/fetcher.go

```diff
@@ -0,0 +1,294 @@
+	"context"
+	"encoding/json"
+	"errors"
+	"io"
+	"math"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/alitto/pond"
+	validator "github.com/asaskevich/govalidator"
+	"github.com/golang/glog"
+	"github.com/patrickmn/go-cache"
+	"github.com/prebid/prebid-server/config"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+var refetchCheckInterval = 300
+
+type FetchInfo struct {
+	config.AccountFloorFetch
+	FetchTime      int64
+	RefetchRequest bool
+}
+
+type WorkerPool interface {
+	TrySubmit(task func()) bool
+	Stop()
+}
+
+type FloorFetcher interface {
+	Fetch(configs config.AccountPriceFloors) (*openrtb_ext.PriceFloorRules, string)
+}
+
+type PriceFloorFetcher struct {
+	pool            WorkerPool      // Goroutines worker pool
+	fetchQueue      FetchQueue      // Priority Queue to fetch floor data
+	fetchInProgress map[string]bool // Map of URL with fetch status
+	configReceiver  chan FetchInfo  // Channel which recieves URLs to be fetched
+	done            chan struct{}   // Channel to close fetcher
+	cache           *cache.Cache    // cache
+	cacheExpiry     time.Duration   // cache expiry time
+}
+
+func (f *PriceFloorFetcher) Fetch(config config.AccountPriceFloors) (*openrtb_ext.PriceFloorRules, string) {
+	if !config.UseDynamicData || len(config.Fetch.URL) == 0 || !validator.IsURL(config.Fetch.URL) {
+		return nil, openrtb_ext.FetchNone
+	}
+
+	// Check for floors JSON in cache
+	result, found := f.Get(config.Fetch.URL)
+	if found {
+		fetcherResult, ok := result.(*openrtb_ext.PriceFloorRules)
+		if !ok || fetcherResult.Data == nil {
+			return nil, openrtb_ext.FetchError
+		}
+		return fetcherResult, openrtb_ext.FetchSuccess
+	}
+
+	//miss: push to channel to fetch and return empty response
+	if config.Enabled && config.Fetch.Enabled && config.Fetch.Timeout > 0 {
+		fetchInfo := FetchInfo{AccountFloorFetch: config.Fetch, FetchTime: time.Now().Unix(), RefetchRequest: false}
+		f.configReceiver <- fetchInfo
+	}
+
+	return nil, openrtb_ext.FetchInprogress
+}
+
+type FetchQueue []*FetchInfo
+
+func (fq FetchQueue) Len() int {
+	return len(fq)
+}
+
+func (fq FetchQueue) Less(i, j int) bool {
+	return fq[i].FetchTime < fq[j].FetchTime
+}
+
+func (fq FetchQueue) Swap(i, j int) {
+	fq[i], fq[j] = fq[j], fq[i]
+}
+
+func (fq *FetchQueue) Push(element interface{}) {
+	fetchInfo := element.(*FetchInfo)
+	*fq = append(*fq, fetchInfo)
+}
+
+func (fq *FetchQueue) Pop() interface{} {
+	old := *fq
+	n := len(old)
+	fetchInfo := old[n-1]
+	old[n-1] = nil
+	*fq = old[0 : n-1]
+	return fetchInfo
+}
+
+func (fq *FetchQueue) Top() *FetchInfo {
+	old := *fq
+	if len(old) == 0 {
+		return nil
+	}
+	return old[0]
+}
+
+func workerPanicHandler(p interface{}) {
+	glog.Errorf("floor fetcher worker panicked: %v", p)
+}
+
+func NewPriceFloorFetcher(maxWorkers, maxCapacity, cacheCleanUpInt, cacheExpiry int) *PriceFloorFetcher {
+	floorFetcher := PriceFloorFetcher{
+		pool:            pond.New(maxWorkers, maxCapacity, pond.PanicHandler(workerPanicHandler)),
```

Comment:
This library was being used due to performance issue if the number of urls are large
if we think of hosting company, then work with around 10k to 20k publisher, let say every publisher using dynamic floor and everyone has atleast 1 url, then then we have atleast 10k urls
if we try to spawn go routines every time we have a request then we need to create 10k go routines and again we need to fetch them periodically, so again we have to spawn it 
it might cause perfomance issues, thats why used pool instead of creating a go routine for every instance
This library takes care if pool is not getting fully utilised it may shrink or enlarge based on the load

Reply:
Let me know if we still want to keep this or switch to normal go-routine per request

Reply:
As decided, we will go with this implementation for now, we will switch to different approach at the time of fetcher 2.0